pkg/backup/file_sst_sink.go

```diff
@@ -331,11 +342,103 @@ func (s *fileSSTSink) write(ctx context.Context, resp exportedSpan) (roachpb.Key
 	return resp.resumeKey, err
 }
 
-// adjustFileEndKey checks if the export respsonse end key can be used as a
-// split point during restore. If the end key is not splitable (i.e. it splits
+func (s *fileSSTSink) writeKey(
+	ctx context.Context,
+	elidedPrefix []byte,
+	key storage.MVCCKey,
+	value []byte,
+	spanEndKey roachpb.Key,
+	start hlc.Timestamp,
+	end hlc.Timestamp,
+) error {
+	if s.out == nil {
+		if err := s.open(ctx); err != nil {
+			return err
+		}
+	}
+	fullKey := key
+	fullKey.Key = append(elidedPrefix, fullKey.Key...)
+	var lastFile *backuppb.BackupManifest_File
+	if len(s.flushedFiles) > 0 {
+		lastFile = &s.flushedFiles[len(s.flushedFiles)-1]
+		lastEndKey := lastFile.Span.EndKey
+		if fullKey.Key.Compare(lastEndKey) < 0 || !bytes.Equal(elidedPrefix, s.elidePrefix) {
+			log.VEventf(
+				ctx, 1, "flushing backup file %s of size %d because key %s cannot append before %s",
+				s.outName, s.flushedSize, fullKey.Key, lastEndKey,
+			)
+			lastFile.Span.EndKey = fullKey.Key
+			s.stats.oooFlushes++
+			if err := s.flushFile(ctx); err != nil {
+				return err
+			}
+		}
+	} else {
+	}
+
+	// See write() for the logic of when to extend a file.
+	var extend bool
+	if s.midRow {
+		extend = true
+	} else if lastFile != nil {
+		extend = lastFile.Span.EndKey.Equal(fullKey.Key) &&
+			lastFile.EndTime == end &&
+			lastFile.StartTime == start &&
+			lastFile.EntryCounts.DataSize < fileSpanByteLimit
+	}
+
+	endRowKey, err := keys.EnsureSafeSplitKey(key.Key)
+	if err != nil {
+		return err
+	}
+	s.midRow = !endRowKey.Equal(key.Key)
+
+	keyAsRowCount := roachpb.RowCount{
+		DataSize: int64(len(fullKey.Key)) + int64(len(value)),
```

Comment:
I think there should be some logic here that allows us to set the `Rows`/`IndexEntries` here (perhaps not `IndexEntries` since that column will be removed from backup/restore output). Is there a way to extract out the index from a key so that we can determine if `Rows` should be incremented?